src/qibolab/instruments/emulator/pulse_simulator.py

```diff
@@ -0,0 +1,758 @@
+                    pulse_signal = i_env * sinalpha + q_env * cosalpha
+                    # pulse_signal = pulse_signal/np.sqrt(2) # uncomment for ibm runcard
+
+                    times_list.append(t)
+                    signals_list.append(pulse_signal)
+
+                    if pulse.type.value == "qd":
+                        platform_channel_name = f"drive-{qubit}"
+                    # TODO: add during flux pulse update
+                    # elif pulse.type.value == "qf":
+                    # platform_channel_name = f"flux-{qubit}"
+                    elif pulse.type.value == "ro":
+                        platform_channel_name = f"readout-{qubit}"
+
+                    emulator_channel_name_list.append(
+                        channel_translator(platform_channel_name, pulse._if)
+                    )
+
+        tmin, tmax = [], []
+        for times in times_list:
+            tmin.append(np.amin(times))
+            tmax.append(np.amax(times))
+
+        tmin = np.amin(tmin)
+        tmax = np.amax(tmax)
+        Nt = int(np.round((tmax - tmin) * sampling_rate * sim_sampling_boost) + 1)
+        full_time_list = np.linspace(tmin, tmax, Nt)
+
+    channel_waveforms = {"time": full_time_list, "channels": {}}
+
+    unique_channel_names = np.unique(emulator_channel_name_list)
+    for channel_name in unique_channel_names:
+        waveform = np.zeros(len(full_time_list))
+
+        for i, pulse_signal in enumerate(signals_list):
+            if emulator_channel_name_list[i] == channel_name:
+                for t_ind, t in enumerate(times_list[i]):
+                    full_t_ind = int(
+                        np.round((t - tmin) * sampling_rate * sim_sampling_boost)
+                    )
+                    waveform[full_t_ind] += pulse_signal[t_ind]
+
+        channel_waveforms["channels"].update({channel_name: waveform})
+
+    return channel_waveforms
+
+
+def apply_readout_noise(
+    samples: dict[Union[str, int], list],
+    readout_error: dict[Union[int, str], list],
+) -> dict[Union[str, int], list]:
+    """Applies readout noise to samples.
+
+    Args:
+        samples (dict): Samples generated from get_samples.
+        readout_error (dict): Dictionary specifying the readout noise for each qubit. Readout noise is specified by a list containing probabilities of prepare 0 measure 1, and prepare 1 measure 0.
+
+    Returns:
+        dict: The noisy sampled qubit values for each qubit labelled by its index.
+
+    Raises:
+        ValueError: If the readout qubits given by samples.keys() is not a subset of the qubits with readout errors specified in readout_error.
+    """
+    # check if readout error is specified for all ro qubits
+    ro_qubit_list = list(samples.keys())
+    if not set(ro_qubit_list).issubset(readout_error.keys()):
+        raise ValueError(f"Not all readout qubits are present in readout_error!")
+    noisy_samples = {}
+    for ro_qubit in ro_qubit_list:  # samples.keys():
```

Comment:
This was previously already reviewed by @alecandido, it's quite far back so it might take some time to find that resolved comment. This condition is there to ensure that readout error is defined for all readout qubits especially when applying the noise separately such as the case demonstrated in the example notebook.

Reply:
I see, I was mainly asking for the comment.
It is needed or we can remove it?

Reply:
Ah yes it can be removed. Sorry it wasn't clear from a single question mark and 4 lines of code highlighted